src/GithubMergeTool/GithubMergeTool.cs

```diff
@@ -163,5 +164,158 @@ git commit
+            // created by other users, but not right now.
+            if ((string)prInfo["user"]["login"] != "dotnet-bot")
+            {
+                return (false, null);
+            }
+
+            // Check that the PR has a well-known 'auto-merge' label
+            var prIssueResponse = await _client.GetAsync($"repos/{repoOwner}/{repoName}/issues/{prId}");
+            if (!prIssueResponse.IsSuccessStatusCode)
+            {
+                return (false, prIssueResponse);
+            }
+
+            var prIssueBody = JObject.Parse(await prIssueResponse.Content.ReadAsStringAsync());
+            if (!prIssueBody["labels"].Any(label => (string)label["name"] == AutoMergeLabelText))
+            {
+                return (false, null);
+            }
+
+            // Check that the PR has no rejections
+            var reviewsResponse = await _client.GetAsync(prUri + "/reviews");
+            if (!reviewsResponse.IsSuccessStatusCode)
+            {
+                return (false, reviewsResponse);
+            }
+
+            var prReviews = JArray.Parse(await reviewsResponse.Content.ReadAsStringAsync());
+            if (!prReviews.All(review => (string)review["state"] != "CHANGES_REQUESTED"))
+            {
+                return (false, reviewsResponse);
+            }
+
+            // Check that there are no failing required tests
+            var mergeBranchRef = (string)prInfo["head"]["ref"];
+            var baseBranchRef = (string)prInfo["base"]["ref"];
+
+            var branchResponse = await _client.GetAsync($"/repos/{repoOwner}/{repoName}/branches/{baseBranchRef}");
+            if (!branchResponse.IsSuccessStatusCode)
+            {
+                return (false, branchResponse);
+            }
+
+            var branchInfo = JObject.Parse(await branchResponse.Content.ReadAsStringAsync());
+            IEnumerable<string> requiredTests = branchInfo["protection"]["required_status_checks"]["contexts"].Values<string>();
+
+            var testStatusResponse = await _client.GetAsync($"repos/{repoOwner}/{repoName}/commits/{mergeBranchRef}/status");
+            if (!testStatusResponse.IsSuccessStatusCode)
+            {
+                return (false, testStatusResponse);
+            }
+
+            var testStatusBody = JObject.Parse(await testStatusResponse.Content.ReadAsStringAsync());
+            var statusDict = testStatusBody["statuses"].Select(t => ((string)t["context"], "success" == (string)t["state"]))
+                .ToDictionary(t => t.Item1, t => t.Item2);
+            foreach (var test in requiredTests)
+            {
+                if (!statusDict[test])
+                {
+                    // There is a failing required test
+                    return (false, null);
+                }
+            }
+
+            // Check if there are any approvals
+            if (!prReviews.Any(review => (string)review["state"] == "APPROVED"))
+            {
+                // If not, mark the PR as approved
```

Comment:
I would argue that approving a PR is beyond the scope of this merge tool:

1. I might want to add the appropriate label to a PR, but I could still be waiting on code review/approval.  E.g., an escrow change.
2.  The account that creates a PR can't be the one that approves it so this will regularly fail.  My suggestion for this is to create another bot _and_ another account that scans for and approves the auto-merge PRs.  That way the busy work of merging is the responsibility of bots, but human-created PRs can still benefit from the auto-merging, but only after getting a real person to approve it.

Reply:
> I might want to add the appropriate label to a PR, but I could still be waiting on code review/approval

Then you should block it using required tests, like marking the PR `[WIP]`.

> The account that creates a PR can't be the one that approves it so this will regularly fail.

This bot never creates PRs, it only approves them. So it will never fail.